### DISCUSSION TITLE
Type 'string' is not assignable to type 'Date'."

### DISCUSSION THREAD

**Discussion by arefay** [prisma/prisma#14371] [Q&A]
I have a simple model which includes a 
`createdAt       DateTime                    @default(now())`
column, generating a SQLite db.

I'm using Remix.run to fetch some rows using a loader function and .findMany(), to then populate some objects, e.g.

`type LoaderData = {
  project: Project;
}`

and convert to json using Remix.run's json function

`return json(data);`

This was all working fine until I updated npm, prisma all to latest and now all my React components show errors related to the "DateTime" column not being converted correctly?

`Types of property 'createdAt' are incompatible.\n    
Type 'string' is not assignable to type 'Date'.",`

Oddly I'm actually not getting any build errors using npm run dev, but i'm getting all the squiggly red lines on all my React components now where I'm trying to pass an object with a DateTime column from my prisma schema.

Has anyone come across this yet? Thanks!

**Comment by austincrim** [comment]
Hey @arefay 👋,

Could you provide a more complete code snippet so I can get the full picture?

**Reply by arefay** [reply]
Sure, so assuming a prisma schema model like so:

```
model User {
id            String                 @id @default(uuid()) 
createdAt     DateTime               @default(now()) 
email         String                 @unique 
}
```

and having a simple Remix.run route page /users.tsx with something like so:

```
import { User } from "@prisma/client";

type LoaderData = {
  users: User[];
};

export const loader: LoaderFunction = async ({ request, params }) => {
  const list = await db.user.findMany({
    take: 20
  });

  const data: LoaderData = {
    users: list,
  };

  return json(data);
};

export default function UsersPage() {
  const data = useLoaderData<LoaderData>();

  return (
    <>
       <UserList users={data.users}/>
   </>
}
```

And a component:

```
import { User } from "@prisma/client";

export default function UserList({users}: {users:User[]}){
    return <>
        {users.map((user)=>{
            return <div>{user.email}</div>
        })}
    </>
}
```

With this setup, VS Code is showing errors related to the UsersPage file in the "Problems" panel saying:
`Types of property 'createdAt' are incompatible.`
  `Type 'string' is not assignable to type 'Date'.`
related to passing {data.users} to the UserList component

Hope this makes it more clear?

**Reply by austincrim** [reply]
That helps! I think the underlying issue is that while the Prisma query returns a `Date` object, instances of `Date` cannot be serialized to JSON and passed from the backend to the frontend. On the frontend, the `createdAt` field will just be a plain string. If you need it to be an instance of `Date`, you will need to re-construct it on the frontend.

As an aside, you might be interested in a [recent Remix release](https://github.com/remix-run/remix/releases/tag/remix%401.6.5) that improves the type-safety of the `useLoaderData` hook.

**Reply by runoncedev** [reply]
Having the exact same problem. Tried the approach suggested on v1.6.5 release changelog but same result.

I have a pretty simple Remix app which reproduces the error:

* https://github.com/oncet/blog/blob/master/app/routes/index.tsx#L39

```
Type 'SerializeObject<{ id: number; title: string; slug: string; image: string; body: string; categoryId: number | null; publishedAt: Date; createdAt: Date; }>' is not assignable to type 'Post'.
  Types of property 'publishedAt' are incompatible.
    Type 'string' is not assignable to type 'Date'.ts(2322)
PostCard.tsx(5, 3): The expected type comes from property 'post' which is declared here on type 'IntrinsicAttributes & PostCardProps & { children?: ReactNode; }'
```

**Reply by runoncedev** [reply]
Just found related issue:

* https://github.com/prisma/prisma/issues/12540

Seems to be the same problem.

I hadn't luck trying to cast the type to `Date` either:

```
post={{
  ...rest,
  publishedAt: post.publishedAt as unknown as Date,
}}
```

Also tried instantiating a new `Date` but same problem:

```
publishedAt: new Date(post.publishedAt)
```

**Reply by austincrim** [reply]
The underlying issue here is the same, Prisma returns `Date`'s but they can't be serialized to JSON. Remix's types are helping here letting us know that what we expect to be a `Date` is actually a string after it comes across the network, denoted by the wrapping `SerializeObject` type:
![CleanShot 2022-08-26 at 10 49 16](https://user-images.githubusercontent.com/32459922/186944016-1d1fdaac-a991-4644-bebf-1cbb4e4506af.png)

The solution is to use something like [superjson](https://www.npmjs.com/package/superjson) or reconstruct the dates on the client side. This worked for me:

```tsx
<PostCard
  post={{
    ...post,
    publishedAt: new Date(post.publishedAt),
    createdAt: new Date(post.createdAt)
  }}
/>
```

**Reply by runoncedev** [reply]
I think I was probably missing the 2nd date attribute. Thank you so much!

**Comment by piotrkulpinski** [comment]
For anyone else having similar issues, setting the type as `SerializeFrom<Type>` fixed the problem with Date fields being incompatible.

```
const Post = ({ post }: { post: SerializeFrom<PostSelectMany> }) => {
  ...
  return null
}

```

**Reply by timrbula** [reply]
Thanks this worked great for me. I may have missed another place it was exported from but I imported `SerializeFrom` from `@remix-run/server-runtime` in case anyone is looking to use this solution.

**Reply by DonKoko** [reply]
This is great. I have been facing this issue a lot and this seems like the best/cleanest solution. Thank you @piotrkulpinski

**Reply by binishjoshi** [reply]
Just a note, `SerailizeFrom` is now available in `@remix-run/node`

Edit: typo below

**Reply by ThaddeusJiang** [reply]
typo:  ~~SerailizeForm~~  -> `SerializeFrom`

```
import { SerializeFrom } from "@remix-run/node"
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
